BPF Map Events Tracking

tommyp1ckles opened the pull request:
Fixes: #8179

Adding event log buffers features, which allows storing of bpf map events in a ring buffers of a specified size. This also includes CLI/API server implementations for fetching ring buffer data via "cilium map events <map_name>".
This will help debug issues where we BPF maps our in an incorrect state, but we don't know why and how they got there.

Example output:

```
$ cilium map events cilium_lxc --follow
key="10.0.1.238:0" value="<nil>" time=2022-11-03T16:56:52Z action=delete desiredState=to-be-deleted lastError="<nil>"
key="[fd00::10a]:0" value="<nil>" time=2022-11-03T16:56:52Z action=delete desiredState=to-be-deleted lastError="<nil>"
key="[fd00::10a]:0" value="<nil>" time=2022-11-03T16:57:07Z action=delete desiredState=to-be-deleted lastError="unable to delete element [fd00::10a]:0 from map cilium_lxc: no such file or directory"
...
```

By default only certain maps will have the buffers enabled, and at a certain size to avoid large memory spikes/usage when fetching event buffer data (i.e. such as during a sysdump bugtool).

As well, this uses a streaming API which returns multiple JSON objects. This is to prevent memory spikes when dumping large buffers (i.e. avoiding copies for json serialization). As well, this facilitates a --follow mode which will stream live updates to CLI.

Other changes:
* Added dumping these buffers with the bugtool dump into json format.

```release-note
Add "cilium map events <map>" command that lists bpf map operation events"
```

tommyp1ckles committed:
pkg/option: add bpf-map-events-buffer configuration agent map option.

This flag is a map type flag that configures per bpf map event buffers
using the following format: cilium_ipcache=true,1024,1h.
Where the tuple represents a boolean for enabling/disabling the event buffer,
the max size of the buffer, and the TTL of the buffer.

Also added default configuration so that certain buffers will be enabled
out of the box.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bpf,pkg/maps: add configuration of bpf.Map's using option flag config.

All relevant maps have events configured based on cli bpf map events
config flat.
Added package singleton getter functions that one-time initialize bpf.Maps upon
first usage.
This solves the problem of cli flag options being processed prior to the map
being created.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
api: add endpoints and API models for map events API.

Added and regenerated OpenAPI API spec to include:
* Endpoint for getting bpf map events.
* API Models for serializing bpf map events.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bugtool: add map events commands for common maps to bugtool.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
container: add generic ring buffer container type for ordered data.

This type is implemented using generic slices. Provides API
for searching by ordered keys and performing "GC" which allows
removal of invalidated entries by shrinking the buffer.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
bpf: add optional map event log buffer to bpf.Map.

Add implementation for ring buffer based map events log.
Map events store a timestamped event log object for every attempted
operation of a bpf.Map.

Map events that do not sync (such as in the case of syscall errors) are
added to the buffer with the relevant sync state.
All subsequent attempts reconcile the bpf map state via the controller
will also be added to the event log buffer.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

squeed commented on api/v1/openapi.yaml:
```diff
@@ -831,7 +831,20 @@ paths:
         '404':
           description: Map not found
-
+  "/map/{name}/events":
```
Are there any limiting parameters we might want to add? A "since" field? Count?

Even cooler would be streaming, of course...

tommyp1ckles replied:
>Even cooler would be streaming, of course... 

Yeah this is a good idea, I was thinking about it initially but decided to go for a simpler approach. But now I'm thinking it may solve some potential issues I've been concerned about such as:

* Marshalling json requires basically a full copy of the buffers elements, we've run into OOM issues during sysdumps before. With streaming we could write these to the socket message by message.
* Updates and events must be 100% serialized, so they're all using the same mutex lock in bpf.Map. This is not a problem when adding events but could become an issue when configuring large buffers (i.e. we need to retain like 24h+ worth of events) which would result in blocking all map operations while the buffer is being dumped.

tommyp1ckles replied:
Update: Reworked this to include streaming, the API no longer serves a single JSON document but rather json delimited "Event" models. This solves the problem stated above since events can be directly encoded/written one-by-one.

squeed commented on pkg/container/ring_buffer_test.go:
```diff
@@ -0,0 +1,206 @@
+package container
```
I'm idly wondering if the new fuzz testing framework would be useful here - could we fuzz buffer operations?

tommyp1ckles replied:
Yeah I think so, I am a bit concerned about the complexity of this buffer implementation being correct - hence all the tests.

squeed commented on pkg/container/ring_buffer.go:
```diff
@@ -0,0 +1,121 @@
+
+// IterateValid calls the callback on each element of the buffer, starting with
+// the first element in the buffer that satisfies "isValid".
+func (eb *RingBuffer[T]) IterateValid(isValid func(T) bool, callback func(T)) {
```
Does it make sense just to call this Filter or Query? And take a generic predicate?

squeed commented on pkg/container/ring_buffer.go:
```diff
@@ -0,0 +1,121 @@
+// IterateValid calls the callback on each element of the buffer, starting with
+// the first element in the buffer that satisfies "isValid".
+func (eb *RingBuffer[T]) IterateValid(isValid func(T) bool, callback func(T)) {
+	if !eb.isFull() {
```
Why does the full-ness of a buffer affect whether or not the predicate is called?

tommyp1ckles replied:
I don't think this is needed

squeed commented on pkg/container/ring_buffer.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func (eb *RingBuffer[T]) at(i int) T {
+	v := eb.buffer[(eb.index+1+i)%len(eb.buffer)]
```
is this the same as `eb.buffer[eb.map(i)]`? And if not, why not?

tommyp1ckles replied:
it is, good catch.

squeed commented on pkg/container/ring_buffer.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func (eb *RingBuffer[T]) firstValidIndex(isValid func(T) bool) int {
+	return sort.Search(len(eb.buffer), func(i int) bool {
```
cute :-)

squeed commented on pkg/container/ring_buffer.go:
```diff
@@ -0,0 +1,121 @@
+// GC clears out invalidated elements in the buffer.
+// This may require copying the entire buffer.
+// It is assumed that if buffer[i] is invalid then every entry [0...i-1] is also not valid.
+func (eb *RingBuffer[T]) GC(isValid func(T) bool) {
```
bikeshed: GC is an overloaded term. maybe Prune or Compact?

tommyp1ckles replied:
Agreed, I like compact ... or cull?

squeed commented on daemon/cmd/map.go:
```diff
@@ -5,13 +5,71 @@ package cmd
+		return restapi.NewGetMapNameNotFound()
+	}
+	mapEvents := []*models.MapEvent{}
+	err := m.DumpEventsWithCallback(func(e *bpf.Event) {
```
Do you have to worry about any racing with goroutines updating the map here?

tommyp1ckles replied:
DumpEventsWithCallback holds the bpf.Maps readlock during the callback.

tommyp1ckles committed:
bpf: add subscribing to bpf.Map events.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
server: provide API server implementation for map events endpoint.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
cli: fix panic while running bpf recorder list.

Switch interface{} list type to correct type to avoid similar type
errors in the future.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

christarazi commented on daemon/cmd/daemon_main.go:
```diff
@@ -983,6 +983,10 @@ func initializeFlags() {
 	flags.Var(option.NewNamedMapOptions(option.APIRateLimitName, &option.Config.APIRateLimit, nil), option.APIRateLimitName, "API rate limiting configuration (example: --rate-limit endpoint-create=rate-limit:10/m,rate-burst:2)")
 	option.BindEnv(Vp, option.APIRateLimitName)
 
+	flags.Var(option.NewNamedMapOptions(option.BPFMapEventBuffers, &option.Config.BPFMapEventBuffers, nil), option.BPFMapEventBuffers, "configuration for BPF map event buffers: (example: --bpf-map-event-buffers cilium_ipcache=enabled:true,size:1024,ttl:1h)")
```
`NewNamedMapOptions()` accepts a validator function to be passed in. I wonder if we can make use of that with the parsing function that you've already implemented.

christarazi commented on api/v1/client/daemon/daemon_client.go:
```diff
@@ -42,6 +43,8 @@ type ClientService interface {
 
 	GetMapName(params *GetMapNameParams) (*GetMapNameOK, error)
 
+	GetMapNameEvents(params *GetMapNameEventsParams, writer io.Writer) (*GetMapNameEventsOK, error)
```
This commit should be squashed with the commit that generated these changes (previous commit).

christarazi commented on pkg/container/ring_buffer.go:
```diff
@@ -0,0 +1,121 @@
+// indexing of ordered data.
+//
+// TODO: Switch this to use generic types for the buffer array once we no
+// longer have to worry about backporting to pre 1.18 versions.
```
Let's move the TODOs into a meta issue for following up on that we don't lose track of them. Referencing the GH issue in the comment can take place of the TODO.

christarazi commented on pkg/bpf/map_linux.go:
```diff
@@ -890,6 +925,18 @@ func (m *Map) Update(key MapKey, value MapValue) error {
 }
 
+// deleteMapEvent is run at every delete map event.
+// If cache is enabled, it will update the cache to reflect the delte.
```
typo

christarazi commented on pkg/bpf/events.go:
```diff
@@ -0,0 +1,147 @@
+					b.buffer.Compact(func(e interface{}) bool {
+						event, ok := e.(*Event)
+						if !ok {
+							panic("BUG: wrong object type in event ring buffer")
```
I'm not sure what the semantics are when we panic instead a controller. For instance, I'm not sure if that will completely kill the controller or if it will keep running. Maybe best to just log an error instead.

christarazi commented on pkg/bpf/events.go:
```diff
@@ -0,0 +1,147 @@
+	eb.buffer.IterateValid(eb.eventIsValid, func(e interface{}) {
+		event, ok := e.(*Event)
+		if !ok {
+			panic("BUG: wrong object type in event ring buffer")
```
Might be best to avoid panicking because we don't want to completely halt the operation of Cilium if we encounter some sort of bug with the event type. BPF map event tracking isn't crucial to the operation of Cilium. We should be able to just continue. Logging an error is probably best here and some mention to report to the developers.

christarazi commented on Documentation/cmdref/cilium_map.md:
```diff
@@ -21,6 +21,7 @@ Access userspace cached content of BPF maps
 ### SEE ALSO
 
 * [cilium](cilium.md)	 - CLI
+* [cilium map events](cilium_map_events.md)	 - Display cached list of events for a BPF map
```
Please squash this with the commit that introduced it

christarazi reviewed (commented):
Overall looks good! Nice work!

christarazi commented on pkg/option/config.go:
```diff
@@ -2231,6 +2235,12 @@ type DaemonConfig struct {
+	// BPFMapEventBuffers has configuration on what BPF map event buffers to enabled
+	// and configuration options for those.
+	BPFMapEventBuffers          map[string]string
+	BPFMapEventBuffersValidator func(val string) (string, error) `json:"-"`
```
Does this need to be exported?

tommyp1ckles replied:
It's set in cmd/daemon_main.go, similar to the FixedIdentityMappingValidator (although it could be argued that neither of these are in the right place).

christarazi commented on pkg/bpf/events.go:
```diff
@@ -0,0 +1,151 @@
+		}
+		callback(event)
+	})
+}
```
`dumpWithCallback` logs an error during the same condition as `eventIsValid`. I think the latter needs to be changed to be like the former.

jibi commented on daemon/cmd/daemon_main.go:
```diff
@@ -987,6 +999,10 @@ func initializeFlags() {
 	flags.Var(option.NewNamedMapOptions(option.APIRateLimitName, &option.Config.APIRateLimit, nil), option.APIRateLimitName, "API rate limiting configuration (example: --rate-limit endpoint-create=rate-limit:10/m,rate-burst:2)")
 	option.BindEnv(Vp, option.APIRateLimitName)
 
+	flags.Var(option.NewNamedMapOptions(option.BPFMapEventBuffers, &option.Config.BPFMapEventBuffers, option.Config.BPFMapEventBuffersValidator), option.BPFMapEventBuffers, "configuration for BPF map event buffers: (example: --bpf-map-event-buffers cilium_ipcache=enabled:true,size:1024,ttl:1h)")
```
Do we need to export this option also on the Helm chart or was it left out on purpose?

Also I think the `--bpf-map-event-buffers cilium_ipcache=enabled:true,size:1024,ttl:1h` example is incorrect, as it seems the option is not expecting `enabled:`, `size:` or `ttl:`?

jibi commented on pkg/defaults/defaults.go:
```diff
@@ -510,3 +510,22 @@ const (
+	// BPFEventBufferConfigs contains default configuration entries for bpf map event buffers.
+	// These are to be merged with the client configuration to create the final config.
+	// Note: The TTL corresponds to GC interval times, which is a somewhat expensive operation.
+	// 		 Under the worst case GC may need to memcopy almost the entire buffer, which will
```
Indentation seems a bit off here

jibi commented on pkg/option/config.go:
```diff
@@ -3928,3 +3946,73 @@ func MightAutoDetectDevices() bool {
+	if c, ok := cs[name]; ok {
+		return c
+	}
+	return BPFEventBufferConfig{
```
I don't think we need to handle this case: if the key is not in the map the lookup will return the `BPFEventBufferConfig` zero value anyway

tommyp1ckles replied:
good point 🤔

jibi commented on pkg/option/config.go:
```diff
@@ -3928,3 +3946,73 @@ func MightAutoDetectDevices() bool {
+// For example: true,100,24h
+// Which refers to enabled=true, maxSize=100, ttl=24hours.
+func ParseEventBufferTupleString(optsStr string) (BPFEventBufferConfig, error) {
+	optsStr = strings.Trim(optsStr, " 	")
```
nit: you can use `strings.TrimSpace` (but I'm not sure why we need this :thinking:)

jibi commented on pkg/option/config.go:
```diff
@@ -3928,3 +3946,73 @@ func MightAutoDetectDevices() bool {
+	if err != nil {
+		return conf, fmt.Errorf("could not parse event buffer ttl duration: %w", err)
+	}
+	if conf.MaxSize < 0 {
```
This should be `size` as `conf.MaxSize` is not yet assigned

jibi commented on pkg/container/ring_buffer.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+func (eb *RingBuffer) incr() {
+	eb.index = (eb.index + 1) % len(eb.buffer)
```
Shouldn't this be modulo `maxSize`?

tommyp1ckles replied:
Until the buffer reaches maxSize, it grows the array.

jibi commented on pkg/container/ring_buffer.go:
```diff
@@ -0,0 +1,118 @@
+// indexing of ordered data.
+type RingBuffer struct {
+	buffer  []interface{}
+	index   int // index of ring buffer head.
```
nit: these should probably be `uint`s

jibi commented on pkg/container/ring_buffer.go:
```diff
@@ -0,0 +1,118 @@
+// maps index in [0:len(buffer)) to the actual index in buffer.
+func (eb *RingBuffer) mapIndex(index int) int {
+	tail := eb.index + 1
+	return (tail + index) % len(eb.buffer)
```
ditto here (`eb.maxSize` instead of `len(eb.buffer)`)

tommyp1ckles replied:
Same as above, the buffer starts with cap=maxSize, but len=0: 

https://github.com/cilium/cilium/pull/21235/files/10ee14fa4165721a0ec4c615acab006e357c8163..1683749e8992ffd61bb0daffa4a88a543bfbb4d5#diff-28690204bd7808f2e104abbdb97b1a7e8a4f460eebb8f8dec62bb0ed1f57bf05R24

jibi commented on pkg/bpf/map_linux.go:
```diff
@@ -238,6 +243,27 @@ func (m *Map) WithCache() *Map {
+// TODO: The IPCache map have many periodic update events added by a controller for entries such as the 0.0.0.0/0 range.
+// These fill the event buffer with possibly unnecessary events.
+// We should either provide an option to aggregate these events, ignore hem from the ipcache event buffer or store them in a separate buffer.
+func (m *Map) WithEvents(enabled bool, size int, ttl time.Duration) *Map {
```
So at the moment we have 2 ways to interact with maps in Cilium: the `bpf` package here, and the `ebpf` one, which is based on `cilium/ebpf` (long term plan is to gradually switch to the latter).

Are you planning to support also the `ebpf` package?

tommyp1ckles replied:
> Are you planning to support also the ebpf package?

Yeah, was focused on ipcache in this one. I think next iteration will include ebpf + non static map names.

jibi commented on pkg/bpf/events.go:
```diff
@@ -20,6 +23,9 @@ const (
 	MapDelete
+	// MapDeleteAll describes a map.DeleteAll event which is aggregated into a single event
+	// to minimize memory and subscription buffer usage.
+	MapDeleteAll
```
I think this belongs to the previous commit?

jibi commented on pkg/maps/ipcache/ipcache.go:
```diff
@@ -162,7 +163,7 @@ type Map struct {
 
 func newIPCacheMap(name string) *bpf.Map {
-	return bpf.NewMap(
+	m := bpf.NewMap(
```
Leftover?

jibi commented on pkg/maps/ipcache/ipcache.go:
```diff
@@ -262,24 +265,34 @@ func (m *Map) supportsDelete() bool {
+	once    = &sync.Once{}
 )
 
+// GetIPCache gets the ipcache Map singleton. If it has not already been done,
```
nit: `IPCacheMap()` perhaps, to keep naming consistent with the other maps?

tommyp1ckles replied:
good idea

sayboras reviewed (approved):
LGTM from cilium/metrics team :100:

christarazi commented on pkg/bpf/events.go:
```diff
@@ -0,0 +1,276 @@
+	event, ok := e.(*Event)
+	if !ok {
+		log.WithError(wrongObjTypeErr(e)).Error("Could not dump contents of events buffer")
+		return true
```
Shouldn't this return false?

christarazi commented on pkg/bpf/events.go:
```diff
@@ -0,0 +1,276 @@
+	var activeSubs []*Handle
+	activeSubsLock := &lock.Mutex{}
+	wg := &sync.WaitGroup{}
+	for i, sub := range eb.subscriptions {
+		if sub.isClosed() { // sub will be removed.
+			continue
+		}
+		wg.Add(1)
+		go func(sub *Handle, i int) {
+			defer wg.Done()
+			if sub.isFull() {
+				err := fmt.Errorf("timed out waiting to send sub map event")
+				log.WithError(err).Warnf("subscription channel buffer %d was full, closing subscription", i)
+				sub.close(err)
+			} else {
+				sub.c <- e
+				activeSubsLock.Lock()
+				activeSubs = append(activeSubs, sub)
+				activeSubsLock.Unlock()
+			}
+		}(sub, i)
+	}
```
How many goroutines could be potentially be spinning up here at once?

tommyp1ckles replied:
Unbounded, but practically it will be bounded by "the number of concurrent clients doing --follow on events"

That's good question though, I could introduce some kind of limit to this, or serialize it into batches. However what are the circumstances under which this would cause problems and does it make sense to implement those limits here. 

I could see this causing issues if someone execed in and spam a bunch of forked `cilium map events ... --follow` but I'm sure many such calls would cause problems under this circumstance. How defensive do we want to be with something like that?

If this is something we're concerned about - I think it makes more sense to add that to the http server impl level by adding some kind of rate limit/max concurrent calls to the endpoint.

lmkwyt

christarazi replied:
> Unbounded, but practically it will be bounded by "the number of concurrent clients doing --follow on events"

:ok_hand: 

I think that's good enough for me IMO. I think leaving a comment saying exactly that before the goroutine code tells future readers that they don't need to necessarily worry about this or if someone comes along and notices this assumption somehow no longer holds due to the expectation that's been explicitly documented. Assumptions / expectations are extremely valuable to document in the code base for our future selves; sometimes it can be difficult to assess whether a snippet of code is problematic, this comment would alleviate that.

Ideally for resource constraints particularly CPU / goroutines, we use something like https://github.com/cilium/workerpool in general to avoid unboundedness. We've seen issues before with memory usage and scheduler saturation in hot spots of the code which don't limit the number of goroutines.

tommyp1ckles replied:
sounds good,

no matter the implementation, if someone starts opening too many follow commands this would cause issues. I think the correct thing would be to just limit the number of concurrent subs allowed at the same time.

tommyp1ckles replied:
actually, I ended up adding a limit on the number of subs. probably much simpler in the long run and it'll be easier to solve any potential issues.

christarazi commented on pkg/container/ring_buffer_test.go:
```diff
@@ -0,0 +1,184 @@
+	buffer.Compact(func(n interface{}) bool {
+		return n.(int) > 95
+	})
+	//assert.Equal([]interface{}{96, 97, 98, 99, 100}, buffer.buffer, "should container everything gt 95")
```
Remove?

christarazi reviewed (commented):
I am still seeing TODOs in the code. As mentioned previously, we should convert them into items in a GH issue and reference the GH issue. This way we can better track what we should do next to improve on this feature.

christarazi commented on pkg/bpf/events.go:
```diff
@@ -0,0 +1,287 @@
+// given a constant buffer size in the case where events arrive at once (i.e. all 100 events
+// arriving at the top of the second).
+//
+// NOTE: Although using timers/timed-contexts seems like an obvious choice for this use case,
+// the timer.After implementation actually uses a large amount of memory. To reduce memory spikes
+// in high throughput cases, we instead just use a sufficiently buffered channel.
```
We have `pkg/inctimer` which I believe resolves the issue with `time.After()`, FWIW.

tommyp1ckles replied:
Thats good to know. I think this approach might still be preferable, timers would also introduce the slight question about how long we want to keep the bpf package waiting as we block on a channel. This approach means that bpf will never block on a client subscription.

squeed commented on cilium/cmd/bpf_recorder_list.go:
```diff
@@ -20,12 +20,12 @@ var bpfRecorderListCmd = &cobra.Command{
 		common.RequireRootPrivilege("cilium bpf recorder list")
-		maps := make([]interface{}, 2)
-		maps[0] = recorder.CaptureMap4
+		maps := make([]recorder.CaptureMap, 2)
```
Any reason not to `make([]recorder.CaptureMap, 0, 2)`? and `append()`?

squeed commented on cilium/cmd/map_event_list.go:
```diff
@@ -0,0 +1,116 @@
+			<-sig
+			cancel()
+			fmt.Fprintf(os.Stderr, "\nReceived an interrupt, disconnecting from event stream...\n\n")
+			os.Exit(0)
```
Do you need this? Admittedly, yes, the whole program is going away, but presumably everything should clean up correctly when the context is cancelled?

tommyp1ckles replied:
The cancel? the handler depends on context to properly unsubscribe from the events watch.

(if it doesn't get cancelled, thats still ok, but it will just keep filling up the buffer it uses for subscription until it hits the limit (currently 32) at which point the sub will be closed and GCd).

squeed commented on pkg/option/config.go:
```diff
@@ -3954,3 +3972,67 @@ func MightAutoDetectDevices() bool {
+
+// ParseVentBufferTupleString parses a event buffer configuration tuple string.
+// For example: true,100,24h
+// Which refers to enabled=true, maxSize=100, ttl=24hours.
```
`enabled` here seems redundant (or confusing). I see a few cases
1. The map is disabled by default, so I'd like to enable it and tweak the parameters. Then just `mapname=100,24`
2. The map is disabled by default, and we'd like to enable it with default size. Then just `mapname,mapname1`
3. The map is enabled by default, and we'd like to disable it. Them `mapname=disabled` would be what I would write, not `mapname=false,0,0` or something.

tommyp1ckles replied:
Did some work on this, I think I prefer consistency for the tuple structure however 👍 on the true->enabled change.

tommyp1ckles committed:
cli: switch to multi line output to improve readability.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
cli: rework output formatting.

Improves previous formatting by replacing tabwriter with using simple
key=data ... formatting.
Tabwriter didn't work well for some map types due to highly variable
data field lengths.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

squeed commented on pkg/option/config.go:
```diff
@@ -3954,3 +3972,68 @@ func MightAutoDetectDevices() bool {
+	enabled := false
+	conf := BPFEventBufferConfig{}
+	if len(opts) != 3 {
+		return conf, fmt.Errorf("unexpected event buffer config value format, should be in format 'mapname=true,100,24h'")
```
nit: error should have `enabled`, not `true`

tklauser replied:
FWIW, I've amended the first commit to fix the error message before merging. No need to re-run CI given it was only a typo fix.

squeed reviewed (approved):
One minor nit, otherwise lgtm.
